base/hashing.jl

```diff
@@ -64,11 +64,79 @@ end
 
 hash(x::QuoteNode, h::UInt) = hash(x.value, hash(QuoteNode, h))
 
-# hashing ranges by component at worst leads to collisions for very similar ranges
-const hashr_seed = UInt === UInt64 ? 0x80707b6821b70087 : 0x21b70087
+## hashing collections ##
+
+const hashaa_seed = UInt === UInt64 ? 0x7f53e68ceb575e76 : 0xeb575e76
+const hashrle_seed = UInt == UInt64 ? 0x2aab8909bfea414c : 0xbfea414c
+function hash{T}(a::AbstractArray{T}, h::UInt)
+    if isleaftype(T)
+        if method_exists(-, (T, T))
+            val = (x1, x2) -> x2 - x1
+        else
+            val = (x1, x2) -> x2
+        end
+    else
+        val = (x1, x2) -> applicable(-, x2, x1) ? x2 - x1 : x2
```

Comment:
I'm just imagining a generic function:

```jl
diffhash(a, b) = b
diffhash(a::Number, b::Number) = a-b
diffhash(a::DateTime, b::DateTime) = a-b
# … etc …
```

Custom types can directly hook into it themselves.  And you don't need to worry about container types since Julia will just do the dispatch on the element types; well-typed containers will automatically be type-stable.